pkg/ingester/series.go

```diff
@@ -63,16 +63,19 @@ func newMemorySeries(m labels.Labels) *memorySeries {
 //
 // The caller must have locked the fingerprint of the series.
 func (s *memorySeries) add(v model.SamplePair) error {
-	// Don't report "no-op appends", i.e. where timestamp and sample
-	// value are the same as for the last append, as they are a
-	// common occurrence when using client-side timestamps
-	// (e.g. Pushgateway or federation).
-	if s.lastSampleValueSet &&
-		v.Timestamp == s.lastTime &&
-		v.Value.Equal(s.lastSampleValue) {
-		return nil
-	}
 	if v.Timestamp == s.lastTime {
+		// Don't report "no-op appends", i.e. where timestamp and sample
+		// value are the same as for the last append, as they are a
+		// common occurrence when using client-side timestamps
+		// (e.g. Pushgateway or federation).
+		if s.lastSampleValueSet && v.Value.Equal(s.lastSampleValue) {
```

Comment:
Swapping the order of these two ifs would allow you to get rid of the `s.lastSampleValueSet` clause.

I guess it could even be collapsed to `if !s.lastSampleValueSet || v.Value.Equal(s.lastSampleValue)` but I think the comments explaining each case are nice.

Reply:
I've done the reordering.